### ISSUE TITLE
Tags from plugins

### ISSUE THREAD

**Issue by Sophist-UK** [rdswift/picard-docs#38]
**Type of change requested**
This is an…
 - [x] Error Report
 - [ ] Addition Request
 - [ ] Deletion Request
 - [ ] Modification Request
 - [ ] Other

**Reason for the change**
I am a little uncomfortable about the Tags & Variables section including descriptions of tags / variables that are created by specific plugins (and by definition not other plugins). I would personally prefer to see a whole separate section about Plugins, perhaps with a simple pointer saying that some plugins create additional tags and variables such as Classical music tags for movements etc.

**Comment by phw** [comment]
I think it is correct to list those tags together with the rest. Those are still tags explicitly supported by Picard, meaning it defines these tags, has mappings to various formats and can present the tag with a proper label. We don't need to hide them. What we could make better is noting on tags that do not get filled by Picard by default to say that.

We could of course add a separate plugins section. What should this include?

**Comment by Sophist-UK** [comment]
1. Any tags (not hidden variables) explicitly supported by Picard (such as replay gain) and which therefore appear correctly in the list of tags should IMO be shown in the normal tags section even if they are only defined by a plugin.

2. Any new tags which are only supported by a plugin, should be listed in a plugins section - which for usefulness and fairness needs to list every v2 plugin in the GitHub repo.

3. Any hidden variables explicitly and only created by a plugin should be listed in the plugins section.

P.S. I don't know if I was having brain ache in August or this change has been made since, but the current doc site has a separate Tags from Plugins page - which pretty much does what this issue is asking for except for completeness (and Genre being attributed to Last.fm).

**Comment by Sophist-UK** [comment]
P.S. I happened to note that ReplayGain tags are not listed in the Tags pages.

We could probably do with a review of the documentation against the Picard code.

**Comment by Sophist-UK** [comment]
P.P.S. I note that the appendices refer to 3 ReplayGain tags that are not defined in the [ReplayGain v2 specification from Hydrogen Audio](http://wiki.hydrogenaud.io/index.php?title=ReplayGain_2.0_specification):

* ReplayGain Album Range
* ReplayGain Reference Loudness
* ReplayGain Track Range

I don't actually have a problem with Picard creating new de-facto standards, but we should try to keep as many people aligned as possible.

**Comment by phw** [comment]
>     1. Any tags (not hidden variables) explicitly supported by Picard (such as replay gain) and which therefore appear correctly in the list of tags should IMO be shown in the normal tags section even if they are only defined by a plugin.

Yes, exactly. That's how it is supposed to be. If a tag is missing it should be added to the docs.

>     2\. Any new tags which are only supported by a plugin, should be listed in a plugins section - which for usefulness and fairness needs to list every v2 plugin in the GitHub repo.

As some plugins have more complex documentation we had some discussion last year about allowing plugins to link to a documentation page. I thought we had a ticket for that, but can't find it now. But the idea was to allow plugins to provide a URL that links to online documentation, which we could link to from within Picard and the website. I am not opposed to have this documentation included in picard-docs if it makes sense. But depends on the plugin really and if the plugin author is willing to use picard-docs for that.

> P.P.S. I note that the appendices refer to 3 ReplayGain tags that are not defined

Those had been added in cooperation with https://github.com/Moonbase59/loudgain

**Comment by rdswift** [comment]
> 1. Any tags (not hidden variables) explicitly supported by Picard (such as replay gain) and which therefore appear correctly in the list of tags should IMO be shown in the normal tags section even if they are only defined by a plugin.

I'm not sure that I understand "explicitly supported by Picard", but okay.

> 2. Any new tags which are only supported by a plugin, should be listed in a plugins section - which for usefulness and fairness needs to list every v2 plugin in the GitHub repo.

This can become a nightmare for maintaining the documentation in this repository, because essentially it would mean including and maintaining the documentation for every listed plugin.  This means someone (likely me) would have to track every change made to every plugin and then update the documentation.  I would much prefer the solution of allowing a plugin's description to provide a link to the full documentation, which can then be maintained by the plugin author.  This is something that I proposed in a [Google Doc](https://docs.google.com/document/d/18Bx5IpKfERnO3Ag529FNiRc_TnU9pLdq7-japRdV3iM/edit) a couple of years ago, which is likely where @phw remembers it from.  At that time @zas was considering making major changes to the plugin system for Picard, and I was trying to gather suggestions for consideration.

> 3. Any hidden variables explicitly and only created by a plugin should be listed in the plugins section.

My preference here is the same as for Item 2.

> P.S. I don't know if I was having brain ache in August or this change has been made since, but the current doc site has a separate Tags from Plugins page - which pretty much does what this issue is asking for except for completeness (and Genre being attributed to Last.fm).

Actually, the ***Tags from Plugins*** page was a holdover from the original Picard documentation on the website, and I believe was intended to convey the understanding that tags can also be provided by plugins and to provide a couple of common examples rather than try to maintain a full list.  In fairness, perhaps all examples should be removed and it should be suggested that the user refer to a plugin's documentation for information about additional tags or variables provided by the plugin.

**Comment by Sophist-UK** [comment]
The obvious place to have documentation for plugins is as a markdown file on the Picard web site.

How easy would it be for the documentation github workflow to check the plugins repo for a list of plugins and to generate a page using that list with links to those plugins that have a Readme.md file in the plugin root directory.

**Comment by rdswift** [comment]
> The obvious place to have documentation for plugins is as a markdown file on the Picard web site.

I'm not sure that I agree, because that removes the maintenance and approval of any documentation changes from the plugin author and transfers it to the people maintaining and approving the Picard website.

> How easy would it be for the documentation github workflow to check the plugins repo for a list of plugins and to generate a page using that list with links to those plugins that have a Readme.md file in the plugin root directory.

I don't know, but I'm guessing that it's more than I have the time and energy to tackle at the moment.  :grimacing:

**Comment by zas** [comment]
> The obvious place to have documentation for plugins is as a markdown file on the Picard web site.

While we could display documentation provided freely by plugin authors, imho this isn't the role of Picard core team to document third-party plugins.

We discussed a lot about future of plugins, and it seems we were more or less ok to decentralize plugin management.
The main idea would be to maintain a list of registered plugins, but to not host their code anymore.
If metadata associated with the plugin provides links to documentation we could actually provide it to Picard's website users.

That said, I guess we'll keep a set of "official" (or rather Picard team-maintained) plugins. From documentation point of view, they shouldn't differ from third-party plugins.

Plugin system is definitively one thing we need to work on soon.

@phw : ^^

**Comment by Sophist-UK** [comment]
> The obvious place to have documentation for plugins is as a markdown file on the Picard web site.

Ooops - I meant to say Picard-Plugins Github Repo.

**Comment by Sophist-UK** [comment]
> While we could display documentation provided freely by plugin authors, imho this isn't the role of Picard core team to document third-party plugins.

I agree. I meant to say Picard-Plugins repo not Picard web site.

> We discussed a lot about future of plugins, and it seems we were more or less ok to decentralize plugin management.
> The main idea would be to maintain a list of registered plugins, but to not host their code anymore.
> If metadata associated with the plugin provides links to documentation we could actually provide it to Picard's website users.

I absolutely do NOT support this approach - for security reasons!!! Picard runs this code without any sandbox to limit what they can do. If you simply link to a plugin source which is not under MBF control, then there is **NOTHING** preventing a malicious actor creating what looks like a benign and useful plugin to get it registered and then to change the code to do something very malicious indeed (IMO most likely acting as a trojan to download other malicious code). By retaining control in our own repo, we are able to scrutinise plugins and both avoid and deter malicious intents.

> That said, I guess we'll keep a set of "official" (or rather Picard team-maintained) plugins. From documentation point of view, they shouldn't differ from third-party plugins.

Really? Which plugins are on this list? And what are the criteria for a plugin being on it?

**Comment by Sophist-UK** [comment]
> > How easy would it be for the documentation github workflow to check the plugins repo for a list of plugins and to generate a page using that list with links to those plugins that have a Readme.md file in the plugin root directory.
> 
> I don't know, but I'm guessing that it's more than I have the time and energy to tackle at the moment. 😬

On reflection, there is already an [automatically generated list of plugins](https://picard.musicbrainz.org/plugins/) - so we could simply link to that.

If we can enhance that automatic generation to include a link to any ReadMe.md file, then that would pretty much satisfy the need for web documentation.

**Comment by zas** [comment]
> I absolutely do NOT support this approach - for security reasons!!! Picard runs this code without any sandbox to limit what they can do. If you simply link to a plugin source which is not under MBF control, then there is **NOTHING** preventing a malicious actor creating what looks like a benign and useful plugin to get it registered and then to change the code to do something very malicious indeed (IMO most likely acting as a trojan to download other malicious code). By retaining control in our own repo, we are able to scrutinise plugins and both avoid and deter malicious intents.

Yes, you're right, this is why we should make it clear to users they run plugins at their own risk.
Your position is that Picard team should review all third-party plugins code and ensure users code is safe, while we all know Picard has no way to run plugins in any kind of sandbox, not sure what's the MetaBrainz Foundation's lawyers position on this.

**Comment by Sophist-UK** [comment]
A. We cannot expect users to review code and check that the plugin is safe. Leaving it entirely to users, particularly it is listed in Picard and installation is a single click away is therefore very dangerous, with a high risk that bad players will jump onto Picard as a means of delivering malware - and therefore a high risk of **severe** damage to the Picard and Musicbrainz/Metabrainz brands.

B. We cannot guarantee that our reviews will avoid malware getting through, but 1) it will act as a major deterrent, 2) malware is likely to be obvious in plugins which are short and generally simple pieces of code so we should catch them almost all of the time and 3) we will have the means to stop it once it is identified. We can still issue a disclaimer saying that the code is 3rd party and whilst we have done our best to ensure that it doesn't contain malware, it is run at their own risk.

C. Downloading a plugin that is NOT listed by Picard is definitely at users' own risk.

Now that we are discussing this, I feel that there should be warnings on Picard's Plugins page, and a bigger warning if they decide to side-load a plugin.

**Comment by Sophist-UK** [comment]
> A. We cannot expect users to review code and check that the plugin is safe. Leaving it entirely to users, particularly it is listed in Picard and installation is a single click away is therefore very dangerous, with a high risk that bad players will jump onto Picard as a means of delivering malware - and therefore a high risk of **severe** damage to the Picard and Musicbrainz/Metabrainz brands.
> 
> B. We cannot guarantee that our reviews will avoid malware getting through, but 1) it will act as a major deterrent, 2) malware is likely to be obvious in plugins which are short and generally simple pieces of code so we should catch them almost all of the time and 3) we will have the means to stop it once it is identified. We can still issue a disclaimer saying that the code is 3rd party and whilst we have done our best to ensure that it doesn't contain malware, it is run at their own risk.
> 
> C. Downloading a plugin that is NOT listed by Picard is definitely at users' own risk.
> 
> Now that we are discussing this, I feel that there should be warnings on Picard's Plugins page, and a bigger warning if they decide to side-load a plugin.

P.S. I will go so far as to state that I will stop using Picard if plugins are not hosted in a MB-owned repo because I consider the likelihood that it will get used to host malware so significant.
